drivers/filters/fltmgr/Context.c

```diff
@@ -43,9 +43,102 @@ NTSTATUS
                      _In_ const FLT_CONTEXT_REGISTRATION *Context)
 {
-    UNREFERENCED_PARAMETER(Filter);
-    UNREFERENCED_PARAMETER(Context);
-    return STATUS_NOT_IMPLEMENTED;
+    PFLT_CONTEXT_REGISTRATION ContextPtr;
+    PALLOCATE_CONTEXT_HEADER ContextHeader, Prev;
+    PVOID Buffer;
+    ULONG BufferSize = 0;
+    USHORT i;
+    NTSTATUS Status;
+
+    /* Loop through all entries in the context registration array */
+    ContextPtr = (PFLT_CONTEXT_REGISTRATION)Context;
+    while (ContextPtr)
+    {
+        /* Bail if we found the terminator */
+        if (ContextPtr->ContextType == FLT_CONTEXT_END)
+            break;
+
+        /* Make sure we have a valid context  */
+        if (IsContextTypeValid(ContextPtr->ContextType))
+        {
+            /* Each context is backed by a crtl struct. Reserve space for it */
+            BufferSize += sizeof(STREAM_LIST_CTRL);
+        }
+
+        /* Move to the next entry */
+        ContextPtr++;
+    }
+
+    /* Bail if we found no valid registration requests */
+    if (BufferSize == 0)
+    {
+        return STATUS_SUCCESS;
+    }
+
+    /* Allocate the pool that'll hold the context crtl structs */
+    Buffer = ExAllocatePoolWithTag(NonPagedPool, BufferSize, FM_TAG_CONTEXT_REGISTA);
+    if (!Buffer) return STATUS_INSUFFICIENT_RESOURCES;
+
+    RtlZeroMemory(Buffer, BufferSize);
+
+    /* Setup our loop data */
+    ContextHeader = Buffer;
+    Prev = NULL;
+    Status = STATUS_SUCCESS;
+
+    for (i = 0; i < MAX_CONTEXT_TYPES; i++)
+    {
+        ContextPtr = (PFLT_CONTEXT_REGISTRATION)Context;
+        while (ContextPtr)
+        {
+            /* We don't support variable sized contents yet */
+            FLT_ASSERT(ContextPtr->Size != FLT_VARIABLE_SIZED_CONTEXTS);
+
+            /* Bail if we found the terminator */
+            if (ContextPtr->ContextType == FLT_CONTEXT_END)
+                break;
+
+            /* Size and pooltag are only checked when ContextAllocateCallback is null */
+            if (ContextPtr->ContextAllocateCallback == FALSE && ContextPtr->PoolTag == FALSE)
+            {
+                Status = STATUS_FLT_INVALID_CONTEXT_REGISTRATION;
+                goto Quit;
+            }
+
+            /* Make sure we have a valid context  */
+            if (IsContextTypeValid(ContextPtr->ContextType))
+            {
+                Status = SetupContextHeader(Filter, ContextPtr, ContextHeader);
+                if (NT_SUCCESS(Status))
+                {
+                    if (Prev)
+                    {
+                        Prev->Next = ContextHeader;
+                    }
+
+                    Filter->SupportedContexts[i] = ContextHeader;
+                }
+            }
+
+            Prev = ContextHeader;
+
+            /* Move to the next entry */
+            ContextPtr++;
+        }
+    }
+
+    if (NT_SUCCESS(Status))
+    {
+        Filter->SupportedContextsListHead = Buffer;
+    }
+    else
+    {
+
+    }
+
+Quit:
+
```

Comment:
Should free Buffer here in the failure case.

Reply:
yeah, I've not quite finished the function yet. I committed it half complete as I broke the PR build :/